### DISCUSSION TITLE
[Share] Create book section from annotation

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-actions-tags#204] [Action Scripts]
### Description

Create book section from annotation. See also #181 

1. Open the book
2. Create an annotation, e.g. highlight
3. Right click on the citation and click `create book section` (or the menu label you assigned)
  <img width="376" alt="image" src="https://github.com/windingwind/zotero-actions-tags/assets/33902321/9402b94f-1391-4825-a401-11fb6745127c">

4. The annotation comment or annotation text will be the `title` of new book section and the link to annotation position will be added to `URL` of the book section.

This is only for annotations right-click menu.

### Event

None

### Operation

Script

### Data

```javascript
/**
 * Create book section from annotation.
 * @author windingwind
 * @usage 
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/204
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/181
 */
if (!item) return;

if (!item.isAnnotation()) return "Not an annotation";

const ZoteroPane = require("ZoteroPane");

const topItem = Zotero.Items.getTopLevel([item])[0];

ZoteroPane.selectItem(topItem.id);

let section;

try {
  section = await ZoteroPane.duplicateAndConvertSelectedItem();
  if (!section) {
    return "Failed to create book section";
  }
} catch (e) {
  return e;
}

section.setField("title", item.annotationComment || item.annotationText || "");

const link = getOpenPDFURL(item);
link && section.setField("url", link);
await section.saveTx();

return "Successfully created book section";

function getOpenPDFURL(annotationItem) {
  const libraryID = annotationItem.libraryID;
  const library = Zotero.Libraries.get(libraryID);
  if (!library) {
    return;
  }
  let link = `zotero://open-pdf/`;
  if (library.libraryType === "user") {
    link += `library/items/${annotationItem.parentItem?.key}?`;
  } else if (library.libraryType === "group") {
    link += `groups/${library.id}/items/${annotationItem.parentItem?.key}?`;
  } else {
    return "";
  }
  let pageIndex = 1;
  try {
    pageIndex = JSON.parse(annotationItem.annotationPosition).pageIndex + 1;
  } catch (e) {
    Zotero.warn(e);
  }
  link += `page=${pageIndex}&annotation=${annotationItem.key}`;
  return link;
}
```

**Comment by much-doge** [comment]
This is awesome! I've been looking for a replacement for Zutilo's ability to create a book section. This works wonder, thanks!

**Comment by blkdark** [comment]
hello. thank you for this. i made a few small changes to your code. I intend for the highlighted text to be used as the title and the comment of the highlighted text to be added to the new item as an author. but it doesn't work. could you help me? this is the code:

/**
 * Create book section from annotation.
 * @param {Object} item - The Zotero annotation item
 * @returns {string} Status message indicating success or failure
 */
if (!item) return "No item provided";
if (!item.isAnnotation()) return "Not an annotation";

const ZoteroPane = require("ZoteroPane");
const topItem = Zotero.Items.getTopLevel([item])[0];
if (!topItem) return "Could not find top level item";

ZoteroPane.selectItem(topItem.id);

let section;
try {
    section = await ZoteroPane.duplicateAndConvertSelectedItem();
    if (!section) {
        return "Failed to create book section";
    }

    // Configurar campos
    section.setField("title", item.annotationText || "");
    const link = getOpenPDFURL(item);
    if (link) {
        section.setField("url", link);
    }

    // Manejar creadores - asegurarnos que creators existe
    if (!section.getCreators()) {
        section.setCreators([]);
    }

    // Si hay un comentario en la anotación, usarlo como nombre del autor
    if (item.annotationComment) {
        const [firstName, ...lastNameParts] = item.annotationComment.split(" ");
        const lastName = lastNameParts.join(" ");
        const newCreator = {
            firstName: firstName || "Unknown",
            lastName: lastName || "Author",
            creatorType: "author"  // Importante: especificar el tipo de creador
        };
        
        // Usar el método apropiado para agregar creadores
        section.setCreators([...section.getCreators(), newCreator]);
    }

    // Guardar los cambios de section primero para asegurar que los creadores se guarden
    await section.saveTx();

    // Añadir tags
    topItem.addTag("Chapter");
    topItem.addTag("BookChapter");
    section.addTag("Chapter");

    // Guardar los cambios restantes
    await Promise.all([
        topItem.saveTx(),
        section.saveTx()
    ]);

    return "Successfully created book section with new author";
} catch (e) {
    console.error("Error creating book section:", e);
    return `Error: ${e.message}`;
}

**Comment by papatistos** [comment]
Is this script (the original one) still working? When I try to execute it via the annotation menu, I get this error:

![CleanShot 2025-01-24 at 12 10 08@2x](https://github.com/user-attachments/assets/c5a93e6c-5891-4808-b95f-b33f0c3d5e45)

I'm not sure what it is complaining about because it did switch to the library pane and the correct item was selected.

**Reply by RK1PF** [reply]
The `duplicateAndConvertSelectedItem` function works only with the item types "book" or "book section".

Remarks: It may sometimes fakely works if you try to use it with items that are not books or book sections.
